IMAP-Science-Operations-Center/sds-data-manager
# Opensearch snapshot

# Change Summary

## Overview
This update adds functionality to backup the OpenSearch database each time a set of files is indexed. The OpenSearch version was also updated to version 2.7 and the CDK version was updated to 2.88. CDK version 2.88 was required to deploy OS 2.7 through the CDK.

## New Dependencies
- requests-aws4auth (used to send authentication through requests)
- requests-mock (used to mock requests in unit tests)
- freezegun (used to keep file timestamp consistent in unit tests)

## New Files
- snapshot.py (mostly copied from Libera. Big thanks to Brian M.)
   - This file includes the code for registering a backup repository (S3 bucket) and taking the OpenSearch snapshot.
   - I think it would be good to update this file in the future to better fit in with the existing opensearch_utils classes, but this should work for now.

## Deleted Files
None

## Updated Files
- sds_data_manager_stack.py
   - Updated OpenSearch to 2.7
   - Added Snapshot Role and policies for Indexer lambda to allow it to take the snapshots
   - Added a new S3 bucket to store snapshots
- indexer.py
   - Added function to take snapshot at end of lambda

## Testing
A new unit test file `test_snapshot.py` was created, `sds_data_manager_stack.py` and `test_open_search_stack.py` were updated. All files are passing all tests.

NOTE: When deploying and testing the snapshots, there are a few extra steps to give proper permissions to OS to take the snapshots. 
1. Log in to your OpenSearch instance dashboard
2. Back in AWS, go to IAM -> Roles and copy your IndexerServiceLamba role's ARN
3. In the OpenSearch dashboard menu, go to Security -> Roles and click on the manage_snapshots role.
4. Go to the Mapped Users tab and click on the Map Users
5. Paste the ARN that you copied into the backed role text bock and click Map

Now when you upload files to the data bucket, a snapshot should be taken. You can confirm those files were created in the snapshot bucket.
